### PR TITLE
YAML Model for Semantic Conventions

### LINKED ISSUE #547: Proposal: Define semantic conventions in JSON/YAML
This is a proposal to define all semantic conventions in a format that can be automatically parse by every language and generate constants without having to manually sync their constant definitions with the markdown text in this repository:
1. Move definitions in opentelemetry-semantic-conventions - languages can use `git submodules` to reference to that repository
1. Specificantion will automatically generate tabels like https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#common-attributes when the `git submodule` is updated.
1. Languages will automatically re-generate all constants when `git submodule` is updated.

This will allow us to keep consistency between languages and consistent description for every semantic convention defined.

Open questions:
1. Which format to use  JSON/YAML/etc.?
1. Does anyone know about a tool to generate the markdown tables from the JSON? A simple tool can be written, but would like to avoid that.
1. Should languages generate the constant code or read the files at runtime?
    * Personally I prefer generated code - fast initialization, compiler optimizations, etc.

### PR DESCRIPTION
**Pull request by thisthat**
**Ready for merge**

# YAML model for Semantic Convention 

This is a proposal to define semantic conventions in a YAML format that resolves #547.

The different language implementations can use `git submodule`s to reference these definitions, so the following can be automatically generated: 

- Constants for semantic attributes  

- Typed Spans (type-safe span builders with named methods/functions rather than having users type or copy-paste strings from the spec doc for attribute keys) 

 For the `opentelemetry-specification` repo, markdown tables will be generated from the YAML files to serve as written documentation. 

## Markdown Tables 

Semantic conventions use Markdown tables to describe the list of attributes that belong to them. 

With a common format, all tables for the semantic conventions will have the same appearance and requirements. It is expected that special HTML tags are used in skeleton/template Markdown files to tag where and which table must be generated, as is with [markdown-toc](https://github.com/jonschlinkert/markdown-toc) that is currently used to generate the TOCs. 

  

  

## Constants for Semantic Attributes 

Currently, for each instrumentation, attribute names must be copied manually from the specification, which is an error-prone activity. Using a common YAML format, languages can automatically generate constant keys for Semantic Attributes, e.g. [JS](https://github.com/open-telemetry/opentelemetry-js/blob/3d9b8228b536c99e09d36df7ac3386e3239f067a/packages/opentelemetry-plugin-http/src/enums/AttributeNames.ts), and [Java](https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/attributes/SemanticAttributes.java). 

  

  

## Typed Spans 

  

Semantic Attributes are an important aspect of OpenTelemetry because they allow to precisely define the information that provides insights into a specific part of a system. Defining the data format using semantic conventions is the only reliable way to allow back ends to reason about the telemetry data they receive and to analyze them further. To support OpenTelemetry users in properly supplying this data, we can provide additional APIs that aid building spans for specific operations by correctly setting the semantic attributes for a given semantic convention. 

For more information, please refer to the prior art in [OpenCensus](https://github.com/census-instrumentation/opencensus-specs/issues/257), [OTEP25](https://github.com/open-telemetry/oteps/pull/25/files), [original specifications](https://github.com/open-telemetry/opentelemetry-specification/pull/14), and the [Java proposal](https://github.com/open-telemetry/opentelemetry-java/pull/964/files). 

Note that, each implementation will define its own API structure for the Typed Spans to use the common patterns offered by the language. 

  

# Description of the model 

  

The fields and their expected values are presented in [allowed_syntax.md](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/allowed_syntax.md). 

  

For a basic attribute description, please refer to the [General Attributes example](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/general.yaml). 

Here, each attribute is specified in a list providing information for the key, which is computed by joining the semantic convention prefix and the attribute name. Furthermore, the type expected for each attribute is defined. In some cases, attributes are relevant to perform sampling decisions. These attributes can be marked as `sampling_relevant`, for example in the gRPC and FaaS semantic conventions the attributes [service](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L9) and [trigger](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/faas.yaml#L11), respectively. Marking an attribute as `sampling_relevant` will help the Typed Spans to request such information before the Span is started so it is available via attributes to the [Sampler](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#sampler). Furthermore, generated Markdown tables will also have a column to indicate if an attribute is sampling relevant or not. 

In the case that an enumeration is expected, the pre-defined enum values are listed and we can additionally set the flag `allow_custom_values` to true to allow other, custom values to be provided. This will help the Typed Span APIs to properly restrict the allowed attribute values while building the span. 

  

In some semantic conventions, a common structure is shared by specialized semantic conventions for specific cases, e.g. [HTTP](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md) and [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md), where a common description is given and then additional attributes are described depending on whether it is a server or client operation. To express this, the `extends` field is added and set to the id of the semantic convention to extend, e.g [gRPC Client](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L17). 

  

Another common case is to reference attributes of another semantic convention. For example, the [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md) semantic convention references many [Network Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/span-general.md#general-network-connection-attributes). To cross-reference attributes between different semantic conventions the `ref` field is used. `ref` also allows overriding the description of the attributes to properly fit it to the current use case. When using `ref`, the constraints that specify if the referenced attribute is required are not taken over, this always has to be explicitly defined for each convention in which it is used. 

An example is the [gRPC Client](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L21). Using `ref` will allow the Markdown table generation to create links between the different semantic conventions. 

  

Semantic Conventions might have special constraints, for example in [gRPC](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/rpc.md) at least one of `net.peer.ip` or `net.peer.port` must be set. This can be expressed in the YAML format using the `any_of` constraint. An example is the [RPC Semantic Convention](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/rpc.yaml#L13). 

Another constraint is that a semantic convention requires that another semantic convention is, in turn, followed as well. For example, in the [FaaS Semantic Convention](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md#http) when a function is used to respond an inbound HTTP request, the [HTTP Server Semantic Convention](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#http-server-semantic-conventions) must also be followed. This can be expressed in the YAML format using the `include` constraint. An example is available [here](https://github.com/dynatrace-oss-contrib/opentelemetry-specification/blob/semconv-model/yaml/span_attribute/faas.yaml#L82). 

# Versioning

Since OpenTelemetry has reached the beta status, it is important to introduce the notion of versioning. Whenever we introduce a semantic change, the version should be bumped by one. This way, backends can correctly ingest OpenTelemetry data.
What we consider a semantic change, is out of the scope of this PR and should be discussed in a separate issue.

### CODE REVIEW

**Comment by bogdandrutu** [comment]
I think this is a very nice PR and well documented, but would like to rise some concerns:
* In this PR we are adding some new properties to the attributes such as `sampling_relevant` and `allow_custom_values`. I would prefer these to be added later as enhancements to the current model.
* We are talking about `Typed Spans` which has not been defined yet.

So I would focus for the moment on not adding new functionality.

File: yaml/span_attribute/faas.yaml [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        type:
+          allow_custom_values: true
+          members:
+            - id: insert
+              value: 'insert'
+            - id: edit
+              value: 'edit'
+            - id: delete
+              value: 'delete'
+        brief: 'Describes the type of the operation that was performed on the data.'
```
Comment (bogdandrutu) [inline yaml/span_attribute/faas.yaml:60]: This is not defined right now, maybe consider to remove it, we should not define new semantic conventions here.

File: yaml/allowed_syntax.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+id    ::= string
+brief ::= string
+note  ::= string
+
+prefix ::= string
+
+# extends MUST point to an existing semconv id
+extends ::= string
+
+span_kind ::= "client"
```
[the remark this answers is not in the document]
Comment (Oberon00) [reply yaml/allowed_syntax.md:25]: However, semantic conventions define which span kind you need to set for certain Spans. E.g. PRODUCER for most messaging senders, SERVER for an HTTP server, etc.

File: yaml/allowed_syntax.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+          |   "consumer"
+          |   "internal"
+
+attributes ::= (id type brief examples | ref [brief] [examples]) [required] [sampling_relevant] [note]
+
+# ref MUST point to an existing attribute id
+ref ::= id
+
+type ::= "string"
+     |   "number"
```
[the remark this answers is not in the document]
Comment (Oberon00) [reply yaml/allowed_syntax.md:37]: This is unfortunately still an unresolved discussion, currently we do indeed have only vaguely defined "numeric" attributes, see #425.

File: yaml/allowed_syntax.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+     |   "boolean"
+     |   "string[]"
+     |   "number[]"
+     |   "boolean[]"
+     |   enum
```
[the remark this answers is not in the document]
Comment (Oberon00) [reply yaml/allowed_syntax.md:42]: Enums are just the YAML's way of saying that only certain string/number values are allowed or to provide a list of constants that are recommended for them. This does not propose a new attribute type.

File: yaml/span_attribute/faas.yaml (hunk 1 above)
Comment (Oberon00) [reply yaml/span_attribute/faas.yaml:60]: What do you mean? I think this is copied straight from https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md#datasource

File: yaml/allowed_syntax.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+allow_custom_values := boolean
+
+members ::= member {member}
+
+member ::= id value [brief] [note]
+
+required ::= "yes"
+         |   "conditional"
+
+sampling_relevant ::= boolean
```
[the remark this answers is not in the document]
Comment (thisthat) [reply yaml/allowed_syntax.md:55]: I agree that this is a new concept. I will remove it and propose it as a new concept in a separate issue.

**Comment by thisthat** [comment]
> I think this is a very nice PR and well documented, but would like to rise some concerns:
> 
> * In this PR we are adding some new properties to the attributes such as `sampling_relevant` and `allow_custom_values`. I would prefer these to be added later as enhancements to the current model.
> * We are talking about `Typed Spans` which has not been defined yet.
> 
> So I would focus for the moment on not adding new functionality.

I do agree with you @bogdandrutu that `sampling_relevant` is a new concept that we should not introduce here. However, `allow_cutom_values` it is not necessarily new. I would argue that currently, it is an implicitly concept (see [`db.type`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md), [`faas.trigger`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/faas.md), [`http.flavor`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md), [`messaging.system`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md), etc) that now is explicit and properly documented.

Regarding your second point, this PR proposes the vision of having Typed Spans. I would argue that this is a (side-effect) benefit we can have for free using the proposed YAML model and not a concept that we are introducing with this PR.

File: yaml/allowed_syntax.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply yaml/allowed_syntax.md:37]: I am +1 to update specs to say explicitly what numeric values we accept

File: yaml/allowed_syntax.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply yaml/allowed_syntax.md:25]: My comment was, do we need to define the values of the SpanKind here ?

File: yaml/allowed_syntax.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply yaml/allowed_syntax.md:42]: Document that. Also what is the difference between `enum` and `allow_custom_values`

File: yaml/allowed_syntax.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply yaml/allowed_syntax.md:55]: Please remove.

**Comment by bogdandrutu** [comment]
> Regarding your second point, this PR proposes the vision of having Typed Spans. I would argue that this is a (side-effect) benefit we can have for free using the proposed YAML model and not a concept that we are introducing with this PR.

Based on the description it seems like the main reason to do this.

**Comment by thisthat** [comment]
@open-telemetry/specs-approvers @open-telemetry/specs-trace-approvers please take another look. 
I add a small example of how we could model with YAML files network and HTTP semantic conventions. I am using these two semantic conventions to showcase how markdown tables can be automatically generated from such definitions. Similarly, I am proposing typed spans in the java-instrumentation repository that are automatically generated from these definitions (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/502).

In this PR, there is also the source code used to generate markdown tables and code. It is not perfect and it requires some more polishing, but we are using it for quite a while internally at Dynatrace. It can be used as a first draft to start avoiding to manually replace tables and code after every semantic convention change (e.g. https://github.com/heyams/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/attributes/SemanticAttributes.java).
The idea is to donate this code to the OpenTelemetry project so everyone can use and improve it.

**Comment by arminru** [comment]
The semantic convention generator which makes use of the YAML model proposed here is available for review in the new build-tools repo: https://github.com/open-telemetry/build-tools/pull/5

**Comment by SergeyKanzhelev** [comment]
@arminru @thisthat It would be great to see this PR adjusted accordingly to use the external tool to minimize the number of files to review.

I'd also suggest to minimize the amount of semantics changes while introducing yaml for semantic conventions. I'd really love to have a yaml declarations merged so we can iterate on it.

**Comment by thisthat** [comment]
@SergeyKanzhelev I agree. My plan is to split in multiple PRs the transition to YAML declarations. I will update this PR proposing only a single semantic convention so we can start iterate on it :)

**Comment by thisthat** [comment]
I updated the PR proposing the semantic convention for http and general attributes. 
I added a Makefile entry to run the docker image that generates the semantic convention tables using the yaml definitions. Furthermore, I created a GitHub action that uses the docker image to verify that the tables are up-to-date with the yaml definitions.
Since https://github.com/open-telemetry/build-tools/pull/5 is not yet approved, I have uploaded the docker image to my personal account and use it in this PR. As soon as the image is published under the OTel account, I will update the required files.

File: yaml/allowed_syntax.md [hunk 6]
```diff
@@ -0,0 +1,235 @@
+# Semantic Convention YAML Language
```
Comment (Oberon00) [inline yaml/allowed_syntax.md:1]: Rename to just syntax.md?

File: specification/trace/semantic_conventions/http.md [hunk 7]
```diff
@@ -189,24 +201,24 @@ If the route does not include the application root, it SHOULD be prepended to th
...
-**[1]**: `http.url` is usually not readily available on the server side but would have to be assembled in a cumbersome and sometimes lossy process from other information (see e.g. <https://github.com/open-telemetry/opentelemetry-python/pull/148>).
-It is thus preferred to supply the raw data that *is* available.
-Namely, one of the following sets is required (in order of usual preference unless for a particular web server/framework it is known that some other set is preferable for some reason; all strings must be non-empty):
+At least one of the following is required:
 
 * `http.scheme`, `http.host`, `http.target`
-* `http.scheme`, `http.server_name`, `net.host.port`, `http.target`
-* `http.scheme`, `net.host.name`, `net.host.port`, `http.target`
+* `http.scheme`, `http.server_name`, [net.host.port](span-general.md), `http.target`
+* `http.scheme`, [net.host.name](span-general.md), [net.host.port](span-general.md), `http.target`
```
Comment (Oberon00) [inline specification/trace/semantic_conventions/http.md:219]: It seems we lost the `code` markup here. The tool should be updated to generate

```suggestion
* `http.scheme`, [`net.host.name`](span-general.md), [`net.host.port`](span-general.md), `http.target`
```

File: specification/trace/semantic_conventions/http.md [hunk 8]
```diff
@@ -104,12 +113,15 @@ For an HTTP client span, `SpanKind` MUST be `Client`.
 If set, `http.url` must be the originally requested URL,
 before any HTTP-redirects that may happen when executing the request.
 
-One of the following sets of attributes is required (in order of usual preference unless for a particular web client/framework it is known that some other set is preferable for some reason; all strings must be non-empty):
+<!-- semconv http.client -->
+
+At least one of the following is required:
```
Comment (Oberon00) [inline specification/trace/semantic_conventions/http.md:118]:
```suggestion
**Additional attribute requirements:** At least one of the following sets of attributes is required:
```

etc

File: Makefile [hunk 9]
```diff
@@ -33,3 +35,8 @@ install-markdown-lint:
 .PHONY: markdown-lint
 markdown-lint:
 	@for f in $(ALL_DOCS); do echo $$f; $(MARKDOWN_LINT) -c .markdownlint.yaml $$f; done
+
+
+.PHONY: table-generation
+table-generation:
+	docker run --rm -v $(PWD)/yaml:/source -v $(PWD)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec
```
Comment (Oberon00) [inline Makefile:42]: I hope this is configured to fail on mismatch? If so, please add a comment and rename the target. I don't think changing the tables from a github action is really possible. The action would need to have commit rights into every PR's repository.

Comment (thisthat) [reply Makefile:42]: This is a Makefile, not the GitHub action. The action only checks if the tables are up-to-date.

File: specification/trace/semantic_conventions/http.md (hunk 7 above)
Comment (thisthat) [reply specification/trace/semantic_conventions/http.md:219]: Yes, I would like to have an agreement on style and wording to use. So I can update the code of the tool to support it.

**Comment by Oberon00** [comment]
We use this YAML already for resources at Dynatrace without issue. I assume metric label(sets) should work just as well, although the tool has no special support for metrics.

File: .github/workflows/semantic-convention-check.yaml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+      - .github/workflows/semantic-convention-check.yml
+      - 'spec/**'
+
+jobs:
+  test-markdown:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - name: Verify that semantic convention tables are up-to-date
+      run: docker run --rm -v $(pwd)/yaml:/source -v $(pwd)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec --md-check
```
Comment (SergeyKanzhelev) [inline .github/workflows/semantic-convention-check.yaml:18]: what is `thisthatdc` here?

File: .github/workflows/semantic-convention-check.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
+name: Semantic Convention Check
+on:
+  push:
+    tags: [ '**' ]
+    branches: [ master ]
+  pull_request:
+    branches: [ master ]
+    paths:
+      - .github/workflows/semantic-convention-check.yml
+      - 'spec/**'
```
Comment (SergeyKanzhelev) [inline .github/workflows/semantic-convention-check.yaml:10]: should it be limited to semantic convention files?

File: yaml/allowed_syntax.md (hunk 6 above)
Comment (SergeyKanzhelev) [inline yaml/allowed_syntax.md:1]: Also a better location would be somewhere under semantic_convention I think? Or name the folder samantic_convention?

File: specification/trace/semantic_conventions/http.md (hunk 7 above)
Comment (SergeyKanzhelev) [reply specification/trace/semantic_conventions/http.md:219]: easiest is to try to make a tool generate something that we used to have here and then update it in a separate PR. Least controversial approach for sure.

File: .github/workflows/semantic-convention-check.yaml (hunk 10 above)
Comment (thisthat) [reply .github/workflows/semantic-convention-check.yaml:18]: That's my private account. As I said in my comment, I will update to `otel` as soon as the other PR gets merged.

File: Makefile (hunk 9 above)
Comment (thisthat) [inline Makefile:42]: I will leave this under my personal account for the moment. This way, people can experiment and try the tool. As soon as the tool is published with the opentelemety account, I will update the make file.

**Review by Oberon00 — APPROVED** [review]
LGTM, but before merging this, the build needs to become green:
* Either https://github.com/open-telemetry/build-tools/pull/5 needs to be merged first, or the GH action needs to be disabled/commented out for now.
* The docfx check needs to be fixed (the `nettransport-attribute` in `span-general.md` anchor is missing):

```
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/span-general.md#L12)Invalid link: '[`net.transport` attribute](#nettransport-attribute)'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/rpc.md#L64)Invalid link: '[net.transport][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/messaging.md#L126)Invalid link: '[`net.transport`][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
[20-08-18 11:34:37.018]Warning:[Postprocess.HandlePostProcessorsWithIncremental.HandlePostProcessors.Processing html.ValidateBookmark](specification/trace/semantic_conventions/database.md#L42)Invalid link: '[`net.transport`][]'. The file specification/trace/semantic_conventions/span-general.md doesn't contain a bookmark named 'nettransport-attribute'.
	4 Warning(s)
```

**Comment by arminru** [comment]
Still pending on https://github.com/open-telemetry/build-tools/pull/5 to be merged (or the build action to be disabled for now).

**Comment by bogdandrutu** [comment]
Rebase?

**Review by arminru — APPROVED** [review]

File: Makefile [hunk 12]
```diff
@@ -34,3 +36,7 @@ install-markdown-lint:
 markdown-lint:
 	@echo $(ALL_DOCS)
 	@for f in $(ALL_DOCS); do echo $$f; $(MARKDOWN_LINT) -c .markdownlint.yaml $$f || exit 1;	done
+
+.PHONY: table-generation
+table-generation:
+	docker run --rm -v $(PWD)/semantic_conventions:/source -v $(PWD)/specification:/spec thisthatdc/semconvgen -f /source markdown -md /spec
```
Comment (arminru) [inline Makefile:42]:
```suggestion
	docker run --rm -v $(PWD)/semantic_conventions:/source -v $(PWD)/specification:/spec otel/semconvgen -f /source markdown -md /spec
```
Should be working now that https://github.com/open-telemetry/build-tools/pull/5 was merged and the image published to https://hub.docker.com/r/otel/semconvgen.

**Comment by SergeyKanzhelev** [comment]
it shows me this strange UI without Update Branch or Merge button:

image

@bogdandrutu does it show the same for you?

**Comment by arminru** [comment]
@SergeyKanzhelev 
@bogdandrutu enabled a setting that all branches have to be brought up to date manually before merging until all PRs picked up some linting fixes he recently added (see #512). I updated the branch manually just now.

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev
> @bogdandrutu enabled a setting that all branches have to be brought up to date manually before merging until all PRs picked up some linting fixes he recently added (see #512). I updated the branch manually just now.

Understood. On other PRs I see a button to update a branch so I can do it myself:

image

Maybe this is because I don't have write permissions on DynaTrace repo to update the branch myself before merging.

**Pull request merged**